src/ddn_struct.jl

```diff
@@ -0,0 +1,235 @@
+"""
+    DDNNode(x::Symbol)
+    DDNNode{x::Symbol}()
+
+Reference to a named node in the POMDP or MDP dynamic decision network (DDN).
+
+Note that `gen(::DDNNode, m, depargs..., rng)` always takes an argument for each dependency whereas `gen(::DDNOut, m, s, a, rng)` only takes `s` and `a` arguments (the inputs to the entire DDN).
+
+`DDNNode` is a "value type". See [the documentation of `Val`](https://docs.julialang.org/en/v1/manual/types/index.html#%22Value-types%22-1) for more conceptual details about value types.
+"""
+struct DDNNode{name} end
+
+DDNNode(name::Symbol) = DDNNode{name}()
+
+"""
+Get the name of a DDNNode.
+"""
+name(::DDNNode{n}) where n = n
+name(::Type{DDNNode{n}}) where n = n
+
+"""
+    DDNOut(x::Symbol)
+    DDNOut{x::Symbol}()
+    DDNOut(::Symbol, ::Symbol,...)
+    DDNOut{x::NTuple{N, Symbol}}()
+
+Reference to one or more named nodes in the POMDP or MDP dynamic decision network (DDN).
+
+Note that `gen(::DDNOut, m, s, a, rng)` always takes `s` and `a` arguments (the inputs to the entire DDN) while `gen(::DDNNode, m, depargs..., rng)` takes a variable number of arguments (one for each dependency).
+
+`DDNOut` is a "value type". See [the documentation of `Val`](https://docs.julialang.org/en/v1/manual/types/index.html#%22Value-types%22-1) for more conceptual details about value types.
+"""
+struct DDNOut{names} end
+
+DDNOut(name::Symbol) = DDNOut{name}()
```

Comment:
Consider making these `Base.@pure`. I just tested this for the issue that @MaximeBouton reported in JuliaPOMDP/RLInterface.jl#25 and it stabilizes the type for Julia 1.0.5. I have never used `Base.@pure` before though. Maybe someone with more experience can tell whether using it here is safe. But from the API docs I'd say it's fine.

Reply:
Which functions should be @pure? Can you just make the change and on this branch directly? Sounds like this might also help the problems you were seeing in our tests

Reply:
I just noticed that Base uses a another macro for the value type implementation:

```
struct Val{x}
end

Val(x) = (@_pure_meta; Val{x}())

macro _pure_meta()
    return Expr(:meta, :pure)
end
```

But I don't understand macros well enough to tell what it does.